packages/cli/src/rollup-plugins/server-components.ts

```diff
@@ -0,0 +1,75 @@
+import { AcornNode, Plugin, SourceDescription } from "rollup";
+import { Program } from "estree";
+import MagicString, { SourceMap } from "magic-string";
+
+export function serverComponentsPlugin({
+  sourceMap,
+}: {
+  sourceMap: boolean;
+}): Plugin {
+  return {
+    name: "server-components",
+
+    transform(code, id) {
+      if (code.includes("'use client'") || code.includes('"use client"')) {
+        const ast: Program = (() => {
+          const babelMeta = this.getModuleInfo(id)!.meta.babel;
+          if (babelMeta?.codeAtBabelTime === code) {
+            return this.getModuleInfo(id)!.meta.babel.ast;
+          }
+          return this.parse(code);
+        })();
+
+        for (const [idx, node] of ast.body.entries()) {
+          if (
+            node.type !== "ExpressionStatement" ||
+            node.expression.type !== "Literal" ||
+            typeof node.expression.value !== "string"
+          ) {
+            return;
+          }
+          if (node.expression.value === "use client") {
+            let magicString = new MagicString(code);
+            this.emitFile({ type: "chunk", id });
+            ast.body.splice(idx, 1);
+            const start: number = (node as any).start;
+            const end: number = (node as any).end;
+            const len = end - start;
+            magicString.overwrite(start, end, " ".repeat(len));
+            let output: SourceDescription = {
+              code: magicString.toString(),
+              ast: (ast as unknown) as AcornNode,
+            };
+
+            if (sourceMap) {
+              output.map = magicString.generateMap({ hires: true });
+            }
+            if (!output.meta) {
+              output.meta = {};
+            }
+            output.meta.isUseClientEntry = true;
```

Comment:
perhaps it would be a good idea to prefix this property with something like `_preconstruct_`?

Reply:
We control all the Rollup plugins used so it doesn't really matter

Reply:
Ah, right - good point. I forgot that this isn't anyhow exposed.